snuba/datasets/discover.py

```diff
@@ -0,0 +1,247 @@
+from abc import ABC
+from datetime import timedelta
+from typing import Mapping, Sequence
+
+from snuba.clickhouse.columns import (
+    Array,
+    ColumnSet,
+    DateTime,
+    FixedString,
+    Nested,
+    Nullable,
+    String,
+    UInt,
+    UUID,
+)
+from snuba.datasets.dataset import TimeSeriesDataset
+from snuba.datasets.dataset_schemas import DatasetSchemas
+from snuba.datasets.factory import get_dataset
+from snuba.datasets.schemas import Schema
+from snuba.query.extensions import QueryExtension
+from snuba.query.parsing import ParsingContext
+from snuba.query.project_extension import ProjectExtension, ProjectWithGroupsProcessor
+from snuba.query.query import Query
+from snuba.query.query_processor import QueryProcessor
+from snuba.query.timeseries import TimeSeriesExtension
+from snuba.request.request_settings import RequestSettings
+from snuba.util import is_condition
+
+EVENTS = 'events'
+TRANSACTIONS = 'transactions'
+
+
+def detect_dataset(query: Query) -> str:
+    """
+    Given a query, we attempt to guess whether it is better to fetch data from the
+    "events" or "transactions" dataset. This is going to be wrong in some cases.
+    """
+    # First check for a top level condition that matches either type = transaction
+    # type != transaction.
+    conditions = query.get_conditions()
+    if conditions:
+        for idx, condition in enumerate(conditions):
+            if is_condition(condition):
+                if tuple(condition) == ('type', '!=', 'transaction'):
+                    return EVENTS
+                elif tuple(condition) == ('type', '=', 'transaction'):
+                    return TRANSACTIONS
+
+    # If there is a condition that references a transactions only field, just switch
+    # to the transactions dataset
+    transaction_column_set = get_dataset('discover') \
+        .get_dataset_schemas() \
+        .get_read_schema() \
+        .get_transactions_only_columns()
```

Comment:
Please pass the column sets this needs in the constructor. While right now it is possible to load a dataset from here, this is a coincidence of the fact that get_dataset loads dependencies inline, but that may not last. Depending on a dataset from here is a circular dependency even if python does notice it.

Reply:
Hmm, yeah. Fixed this up.